### PR TITLE
Mixpanel Generate Product Purchased events from Order Completed events

### PR DESCRIPTION
**Pull request by jaredmixpanel**
This PR adds the ability to generate a `Product Purchased` event for every item in the `products` array property of an `Order Completed` event.  This PR will also add an `$anon_id` and `$identified_id` property to every tracked event for optimal compatibility with Mixpanel's identity management system. 

## Testing

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+    description:
+      'The timestamp of the event. If time is not sent with the event, it will be set to the time our servers receive it.',
+    default: {
+      '@path': '$.timestamp'
+    }
+  },
+  app_name: {
+    label: 'App Name',
+    type: 'string',
+    description: 'The name of your application',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:45]:
```suggestion
    description: 'The name of your application.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    type: 'string',
+    description: 'A unique identifier for the device the user is using.',
+    default: {
+      '@path': '$.context.device.id'
+    }
+  },
+  device_type: {
+    label: 'Device Type',
+    type: 'string',
+    description: "The type of the user's device",
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:101]:
```suggestion
    description: "The type of the user's device.",
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    default: {
+      '@path': '$.context.device.type'
+    }
+  },
+  device_name: {
+    label: 'Device Name',
+    type: 'string',
+    description: "The name of the user's device",
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:109]:
```suggestion
    description: "The name of the user's device.",
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    type: 'string',
+    description: 'The device model that the user is using.',
+    default: {
+      '@path': '$.context.device.model'
+    }
+  },
+  bluetooth: {
+    label: 'Bluetooth Enabled',
+    type: 'boolean',
+    description: 'Whether bluetooth is enabled',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:133]:
```suggestion
    description: 'Whether bluetooth is enabled.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+    type: 'string',
+    description: 'The carrier that the user is using.',
+    default: {
+      '@path': '$.context.network.carrier'
+    }
+  },
+  cellular: {
+    label: 'Cellular Enabled',
+    type: 'boolean',
+    description: 'Whether cellular is enabled',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:149]:
```suggestion
    description: 'Whether cellular is enabled.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+    type: 'string',
+    description: 'The language set by the user.',
+    default: {
+      '@path': '$.context.locale'
+    }
+  },
+  library_name: {
+    label: 'Library Name',
+    type: 'string',
+    description: 'The name of the SDK used to send events',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:189]:
```suggestion
    description: 'The name of the SDK used to send events.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+    default: {
+      '@path': '$.context.library.name'
+    }
+  },
+  library_version: {
+    label: 'Library Version',
+    type: 'string',
+    description: 'The version of the SDK used to send events',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:197]:
```suggestion
    description: 'The version of the SDK used to send events.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+    type: 'number',
+    description: 'Pixel density of the device screen.',
+    default: {
+      '@path': '$.context.screen.density'
+    }
+  },
+  referrer: {
+    label: 'Referrer',
+    type: 'string',
+    description: 'Referrer url',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:257]:
```suggestion
    description: 'Referrer url.',
```

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    label: 'Batch Data to Mixpanel',
+    description: 'Set as true to ensure Segment sends data to Mixpanel in batches.',
+    default: true
+  }
+}
+
+export const productsProperties: Record<string, InputField> = {
+  products: {
+    label: 'Products',
+    description: 'Products in the order',
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:331]:
```suggestion
    description: 'Products in the order.',
```

Comment (LLuque-twilio) [reply packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts]: I see there are no periods at the end of the description fields for the `productsProperties` here. I would add them for each to remain consistent with your `eventProperties` descriptions.

File: packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+  },
+  enable_batching: {
+    type: 'boolean',
+    label: 'Batch Data to Mixpanel',
+    description: 'Set as true to ensure Segment sends data to Mixpanel in batches.',
+    default: true
+  }
+}
+
+export const productsProperties: Record<string, InputField> = {
```
Comment (LLuque-twilio) [inline packages/destination-actions/src/destinations/mixpanel/mixpanel-properties.ts:328]: I recommend adding periods to the end of each description field in your `productsProperties` to remain consistent with your `eventProperties` descriptions.

**Review by LLuque-twilio — COMMENTED** [review]
Our partner engineer on rotation will take a more in-depth look at this PR later this week. Just leaving a few nits in the meantime!

File: packages/destination-actions/src/destinations/mixpanel/trackPurchase/__tests__/index.test.ts [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    }
+    const responses = await testDestination.testAction('trackPurchase', {
+      event,
+      mapping,
+      useDefaultMappings: true,
+      settings: settingsObj
+    })
+    expect(responses.length).toBe(1)
+    expect(responses[0].status).toBe(200)
+    expect(responses[0].options.body).toMatchInlineSnapshot(
```
Comment (SyedWasiHaider) [inline packages/destination-actions/src/destinations/mixpanel/trackPurchase/__tests__/index.test.ts:221]: Minor nit but kind of hard to reason about this with the snapshot. I can't tell just by looking at this that that 2 events were created.

File: packages/destination-actions/src/destinations/mixpanel/trackPurchase/index.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    let time = datetime && dayjs.utc(datetime).isValid() ? dayjs.utc(datetime).valueOf() : Date.now()
+    const insertId = payload.insert_id ?? cheapGuid()
+    let insertIdCount = 0
+    purchaseEvents = payload.products.map((product) => {
+      return {
+        event: "Product Purchased",
+        properties: {
+          ...omit(orderCompletedEvent.properties, topLevelKeys),
+          $insert_id: insertId + (insertIdCount++).toString(),
+          time: --time,
```
Comment (SyedWasiHaider) [inline packages/destination-actions/src/destinations/mixpanel/trackPurchase/index.ts:32]: Any reason for the `--` here?

**Review by SyedWasiHaider — APPROVED** [review]
LGTM. Some very minor non-blocking comments/questions.

File: packages/destination-actions/src/destinations/mixpanel/trackPurchase/index.ts (hunk 12 above)
Comment (zihejia) [reply packages/destination-actions/src/destinations/mixpanel/trackPurchase/index.ts:32]: hi @SyedWasiHaider , for each Product Purchased event we generate, 1ms will be decremented from the `Order Completed` time and used as the timestamp. So the events happen in chronological order from Product Purchased 0, Product Purchased 1… Order Completed.

**Pull request merged**